Mathlib/CategoryTheory/Triangulated/Adjunction.lean

```diff
@@ -0,0 +1,143 @@
+import Mathlib.CategoryTheory.Shift.Adjunction
+import Mathlib.CategoryTheory.Adjunction.Additive
+
+/-!
+# The adjoint is triangulated
+
+If a functor `F : C ⥤ D` between pretriangulated categories is triangulated, and if we
+have an adjunction `F ⊣ G`, then `G` is also a triangulated functor.
+
+We deduce that, if `E : C ≌ D` is an equivalence of pretriangulated categories, then
+`E.functor` is triangulated if and only if `E.inverse` is triangulated.
+
+TODO: The case of left adjoints.
+-/
+
+namespace CategoryTheory
+
+open Category Limits Preadditive Pretriangulated Adjunction
+
+variable {C D : Type*} [Category C] [Category D] [HasZeroObject C] [HasZeroObject D]
+  [Preadditive C] [Preadditive D] [HasShift C ℤ] [HasShift D ℤ]
+  [∀ (n : ℤ), (shiftFunctor C n).Additive] [∀ (n : ℤ), (shiftFunctor D n).Additive]
+  [Pretriangulated C] [Pretriangulated D]
+
+namespace Adjunction
+
+variable {F : C ⥤ D} {G : D ⥤ C} (adj : F ⊣ G) [F.CommShift ℤ] [G.CommShift ℤ]
+  [adj.CommShift ℤ] [F.IsTriangulated]
+
+include adj in
+lemma isTriangulated_rightAdjoint : G.IsTriangulated where
+  map_distinguished T hT := by
+    have : G.Additive := adj.right_adjoint_additive
+    dsimp
+    obtain ⟨Z, f, g, mem⟩ := distinguished_cocone_triangle (G.map T.mor₁)
+    obtain ⟨h, ⟨h₁, h₂⟩⟩ := complete_distinguished_triangle_morphism _ _
+      (F.map_distinguished _ mem) hT (adj.counit.app T.obj₁) (adj.counit.app T.obj₂) (by simp)
+    dsimp at h h₁ h₂
+    have h₁' : f ≫ adj.unit.app Z ≫ G.map h = G.map T.mor₂ := by
+      simpa [homEquiv_apply] using congr_arg (adj.homEquiv _ _).toFun h₁
+    have h₂' : g ≫ (G.commShiftIso (1 : ℤ)).inv.app T.obj₁ =
+        (adj.homEquiv _ _ h) ≫ G.map T.mor₃ := by
+      apply (adj.homEquiv _ _).symm.injective
+      simp only [Functor.comp_obj, homEquiv_counit, Functor.id_obj, Functor.map_comp, assoc,
+        homEquiv_unit, counit_naturality, counit_naturality_assoc, left_triangle_components_assoc,
+        ← h₂, adj.shift_counit_app, Iso.hom_inv_id_app_assoc]
+    rw [assoc] at h₂
+    have : Mono (adj.homEquiv _ _ h) := by
+      rw [mono_iff_cancel_zero]
+      intro Y φ hφ
+      obtain ⟨ψ, rfl⟩ := Triangle.coyoneda_exact₃ _ mem φ (by
+        dsimp
+        simp [homEquiv_unit] at hφ
+        rw [← cancel_mono ((G.commShiftIso (1 : ℤ)).inv.app T.obj₁), assoc, h₂', zero_comp,
+          homEquiv_unit, assoc, reassoc_of% hφ, zero_comp])
+      dsimp at ψ hφ ⊢
+      obtain ⟨α, hα⟩ := T.coyoneda_exact₂ hT ((adj.homEquiv _ _).symm ψ)
+        ((adj.homEquiv _ _).injective (by simpa [homEquiv_counit, homEquiv_unit, ← h₁'] using hφ))
+      have eq := congr_arg (adj.homEquiv _ _ ).toFun hα
+      simp only [homEquiv_counit, Functor.id_obj, Equiv.toFun_as_coe, homEquiv_unit,
+        Functor.comp_obj, Functor.map_comp, unit_naturality_assoc, right_triangle_components,
+        comp_id] at eq
+      rw [eq, assoc, assoc]
+      erw [comp_distTriang_mor_zero₁₂ _ mem, comp_zero, comp_zero]
```

Comment:
It would be nice to remove this `erw`.

Reply:
I remove it in the stupidest possible way.